Index: gtk+-2.17.11/configure.in
===================================================================
--- gtk+-2.17.11.orig/configure.in	2009-09-07 12:53:36.000000000 +1000
+++ gtk+-2.17.11/configure.in	2009-09-07 12:53:44.000000000 +1000
@@ -1783,7 +1783,7 @@
 CAIRO_PREFIX="`pkg-config --variable=prefix cairo`"
 
 if test $enable_explicit_deps != yes ; then
-  GDK_PIXBUF_EXTRA_LIBS=
+  GDK_PIXBUF_EXTRA_LIBS="$MATH_LIB"
 fi
 
 AC_SUBST(GDK_PIXBUF_PACKAGES)
