Index: gtk+-2.17.9/configure.in
===================================================================
--- gtk+-2.17.9.orig/configure.in	2009-08-27 16:21:45.000000000 +1000
+++ gtk+-2.17.9/configure.in	2009-08-27 16:21:59.000000000 +1000
@@ -1782,7 +1782,7 @@
 CAIRO_PREFIX="`pkg-config --variable=prefix cairo`"
 
 if test $enable_explicit_deps != yes ; then
-  GDK_PIXBUF_EXTRA_LIBS=
+  GDK_PIXBUF_EXTRA_LIBS="$MATH_LIB"
 fi
 
 AC_SUBST(GDK_PIXBUF_PACKAGES)
