=== modified file 'gtk/gtk.symbols'
Index: gtk+-2.23.2/gtk/gtk.symbols
===================================================================
--- gtk+-2.23.2.orig/gtk/gtk.symbols	2010-11-11 17:52:13.134619001 +1100
+++ gtk+-2.23.2/gtk/gtk.symbols	2010-11-11 17:52:18.462619001 +1100
@@ -5315,6 +5315,7 @@
 gtk_widget_get_realized
 gtk_widget_set_mapped
 gtk_widget_get_mapped
+ubuntu_gtk_widget_set_has_grab
 #endif
 #endif
 
Index: gtk+-2.23.2/gtk/gtkwidget.c
===================================================================
--- gtk+-2.23.2.orig/gtk/gtkwidget.c	2010-11-11 17:52:06.242619001 +1100
+++ gtk+-2.23.2/gtk/gtkwidget.c	2010-11-11 17:52:18.466619001 +1100
@@ -11410,5 +11410,15 @@
   return res;
 }
 
+void
+ubuntu_gtk_widget_set_has_grab (GtkWidget *widget,
+                                gboolean   has_grab)
+{
+  if (has_grab)
+    GTK_OBJECT_FLAGS (widget) |= GTK_HAS_GRAB;
+  else
+    GTK_OBJECT_FLAGS (widget) &= ~(GTK_HAS_GRAB);
+}
+
 #define __GTK_WIDGET_C__
 #include "gtkaliasdef.c"
Index: gtk+-2.23.2/gtk/gtkwidget.h
===================================================================
--- gtk+-2.23.2.orig/gtk/gtkwidget.h	2010-11-11 17:52:06.130619001 +1100
+++ gtk+-2.23.2/gtk/gtkwidget.h	2010-11-11 17:52:18.466619001 +1100
@@ -1344,6 +1344,9 @@
 						       GtkWidget *toplevel,
 						       gpointer   user_data);
 
+void   ubuntu_gtk_widget_set_has_grab  (GtkWidget *widget,
+                                        gboolean   has_grab);
+
 G_END_DECLS
 
 #endif /* __GTK_WIDGET_H__ */
