Index: gtk+-2.23.3/gtk/gtktreeview.c
===================================================================
--- gtk+-2.23.3.orig/gtk/gtktreeview.c	2010-12-22 13:34:53.000000000 +1100
+++ gtk+-2.23.3/gtk/gtktreeview.c	2010-12-23 14:37:50.746944001 +1100
@@ -9622,7 +9622,8 @@
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
