Index: gtk+-2.17.11/gtk/gtktreeview.c
===================================================================
--- gtk+-2.17.11.orig/gtk/gtktreeview.c	2009-09-05 01:50:22.000000000 +1000
+++ gtk+-2.17.11/gtk/gtktreeview.c	2009-09-07 12:53:53.000000000 +1000
@@ -9492,7 +9492,8 @@
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
