Index: gtk+-2.17.9/docs/reference/gtk/Makefile.am
===================================================================
--- gtk+-2.17.9.orig/docs/reference/gtk/Makefile.am	2009-08-07 13:42:22.000000000 +1000
+++ gtk+-2.17.9/docs/reference/gtk/Makefile.am	2009-08-27 16:22:27.000000000 +1000
@@ -363,17 +363,17 @@
 
 ########################################################################
 
-if ENABLE_MAN
-
 man_MANS = gtk-query-immodules-2.0.1 gtk-update-icon-cache.1 gtk-builder-convert.1
 
+if ENABLE_MAN
+
 %.1 : %.xml 
 	@XSLTPROC@ -nonet http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl $<
 
-BUILT_EXTRA_DIST = $(man_MANS)
-
 endif
 
+BUILT_EXTRA_DIST = $(man_MANS)
+
 dist-hook-local: $(BUILT_EXTRA_DIST)
 	files='$(BUILT_EXTRA_DIST)';				\
 	for f in $$files; do					\
Index: gtk+-2.17.9/docs/reference/gdk-pixbuf/Makefile.am
===================================================================
--- gtk+-2.17.9.orig/docs/reference/gdk-pixbuf/Makefile.am	2009-07-20 12:33:46.000000000 +1000
+++ gtk+-2.17.9/docs/reference/gdk-pixbuf/Makefile.am	2009-08-27 16:22:27.000000000 +1000
@@ -79,17 +79,17 @@
 
 ########################################################################
 
-if ENABLE_MAN
-
 man_MANS = gdk-pixbuf-csource.1 gdk-pixbuf-query-loaders.1 
 
+if ENABLE_MAN
+
 %.1 : %.xml 
 	@XSLTPROC@ -nonet http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl $<
 
-BUILT_EXTRA_DIST = $(man_MANS)
-
 endif
 
+BUILT_EXTRA_DIST = $(man_MANS)
+
 dist-hook-local: $(BUILT_EXTRA_DIST)
 	files='$(BUILT_EXTRA_DIST)';				\
 	for f in $$files; do					\
