Index: gtk+2.0-2.19.7/gdk-pixbuf/Makefile.am
===================================================================
--- gtk+2.0-2.19.7.orig/gdk-pixbuf/Makefile.am	2010-03-22 17:21:12.362950828 +0100
+++ gtk+2.0-2.19.7/gdk-pixbuf/Makefile.am	2010-03-22 17:21:12.502966952 +0100
@@ -692,6 +692,7 @@
 
 if HAVE_INTROSPECTION
 
+GdkPixbuf-2.0.gir: libgdk_pixbuf-2.0.la Makefile
 GdkPixbuf_2_0_gir_SCANNERFLAGS = --strip-prefix=Gdk
 GdkPixbuf_2_0_gir_INCLUDES = GModule-2.0 Gio-2.0
 GdkPixbuf_2_0_gir_LIBS = libgdk_pixbuf-2.0.la
Index: gtk+2.0-2.19.7/gdk/Makefile.am
===================================================================
--- gtk+2.0-2.19.7.orig/gdk/Makefile.am	2010-03-22 17:21:30.003951852 +0100
+++ gtk+2.0-2.19.7/gdk/Makefile.am	2010-03-22 17:22:13.358952833 +0100
@@ -201,6 +201,7 @@
         gdkenumtypes.h \
         $(srcdir)/x11/*.c
 
+Gdk-2.0.gir: $(gdktargetlib) Makefile
 Gdk_2_0_gir_SCANNERFLAGS = --strip-prefix=Gdk --add-include-path=../gdk-pixbuf
 Gdk_2_0_gir_INCLUDES = Gio-2.0 GdkPixbuf-2.0 Pango-1.0
 Gdk_2_0_gir_LIBS = $(gdktargetlib)
