Index: gtk+2.0-2.20.0/gdk/Makefile.am
===================================================================
--- gtk+2.0-2.20.0.orig/gdk/Makefile.am	2010-03-22 17:52:46.000000000 +0100
+++ gtk+2.0-2.20.0/gdk/Makefile.am	2010-03-24 11:32:30.145645086 +0100
@@ -199,7 +199,7 @@
 		$(gdk_c_sources) \
         gdkenumtypes.c \
         gdkenumtypes.h \
-        $(srcdir)/x11/*.c
+        x11/*.c
 
 Gdk-2.0.gir: $(gdktargetlib) Makefile
 Gdk_2_0_gir_SCANNERFLAGS = --strip-prefix=Gdk --add-include-path=../gdk-pixbuf
