Index: gtk+2.0-2.21.2/gtk/gtkcontainer.c
===================================================================
--- gtk+2.0-2.21.2.orig/gtk/gtkcontainer.c	2010-06-10 14:53:46.000000000 +0200
+++ gtk+2.0-2.21.2/gtk/gtkcontainer.c	2010-06-29 11:02:30.000000000 +0200
@@ -48,6 +48,8 @@
   REMOVE,
   CHECK_RESIZE,
   SET_FOCUS_CHILD,
+  CHILD_ADDED,
+  CHILD_REMOVED,
   LAST_SIGNAL
 };
 
@@ -291,6 +293,24 @@
 		  _gtk_marshal_VOID__OBJECT,
 		  G_TYPE_NONE, 1,
 		  GTK_TYPE_WIDGET);
+  container_signals[CHILD_ADDED] =
+    g_signal_new (I_("child-added"),
+                  G_OBJECT_CLASS_TYPE (object_class),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL, NULL,
+                  _gtk_marshal_VOID__OBJECT,
+                  G_TYPE_NONE, 1,
+                  GTK_TYPE_WIDGET);
+  container_signals[CHILD_REMOVED] =
+    g_signal_new (I_("child-removed"),
+                  G_OBJECT_CLASS_TYPE (object_class),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL, NULL,
+                  _gtk_marshal_VOID__OBJECT,
+                  G_TYPE_NONE, 1,
+                  GTK_TYPE_WIDGET);
 }
 
 static void
Index: gtk+2.0-2.21.2/gtk/gtkmenushell.c
===================================================================
--- gtk+2.0-2.21.2.orig/gtk/gtkmenushell.c	2010-06-29 11:02:35.000000000 +0200
+++ gtk+2.0-2.21.2/gtk/gtkmenushell.c	2010-06-29 11:02:58.000000000 +0200
@@ -533,6 +533,8 @@
                               menu_shell,
                               child,
                               position);
+
+  g_signal_emit_by_name (menu_shell, "child-added", child);
 }
 
 static void
