From f9a3ce06cc990c8ae3a9d6cee514d0c5f5f05fd8 Mon Sep 17 00:00:00 2001
From: Christian Dywan <christian@twotoasts.de>
Date: Fri, 12 Feb 2010 18:10:17 +0100
Subject: [PATCH] Remove duplicate definitions of gtk_widget_set/get_realized/mapped

---
 gtk/gtkwidget.h |    6 ------
 1 files changed, 0 insertions(+), 6 deletions(-)

diff --git a/gtk/gtkwidget.h b/gtk/gtkwidget.h
index 2490d8b..757be3f 100644
--- a/gtk/gtkwidget.h
+++ b/gtk/gtkwidget.h
@@ -874,12 +874,6 @@ void	gtk_widget_child_notify		  (GtkWidget	       *widget,
 					   const gchar	       *child_property);
 void	gtk_widget_thaw_child_notify	  (GtkWidget	       *widget);
 
-void       gtk_widget_set_realized        (GtkWidget           *widget,
-                                           gboolean             realized);
-gboolean   gtk_widget_get_realized        (GtkWidget           *widget);
-void       gtk_widget_set_mapped          (GtkWidget           *widget,
-                                           gboolean             mapped);
-gboolean   gtk_widget_get_mapped          (GtkWidget           *widget);
 void       gtk_widget_set_can_focus       (GtkWidget           *widget,
                                            gboolean             can_focus);
 gboolean   gtk_widget_get_can_focus       (GtkWidget           *widget);
-- 
1.6.6.1

