From ba3e997daea267b100988ecd7c6d0767c334dfbc Mon Sep 17 00:00:00 2001
From: Benjamin Otte <otte@redhat.com>
Date: Mon, 03 May 2010 16:22:42 +0000
Subject: Make toolbutton not call functions on non-GtkMisc

---
diff --git a/gtk/gtktoolbutton.c b/gtk/gtktoolbutton.c
index 4225401..0e6ae59 100644
--- a/gtk/gtktoolbutton.c
+++ b/gtk/gtktoolbutton.c
@@ -470,11 +470,11 @@ gtk_tool_button_construct_contents (GtkToolItem *tool_item)
 	  gtk_widget_show (icon);
 	}
 
-      if (icon && text_orientation == GTK_ORIENTATION_HORIZONTAL)
+      if (GTK_IS_MISC (icon) && text_orientation == GTK_ORIENTATION_HORIZONTAL)
 	gtk_misc_set_alignment (GTK_MISC (icon),
 				1.0 - gtk_tool_item_get_text_alignment (GTK_TOOL_ITEM (button)),
 				0.5);
-      else if (icon)
+      else if (GTK_IS_MISC (icon))
 	gtk_misc_set_alignment (GTK_MISC (icon),
 				0.5,
 				gtk_tool_item_get_text_alignment (GTK_TOOL_ITEM (button)));
--
cgit v0.8.3.1

